Objects/complexobject.c

```diff
@@ -526,7 +526,10 @@ complex_pow(PyObject *v, PyObject *w, PyObject *z)
     }
     errno = 0;
     exponent = b;
-    int_exponent = (long)exponent.real;
+    if ((double)LONG_MIN <= exponent.real && exponent.real <= (double)LONG_MAX)
```

Comment:
if the sanitizer is happy with this, i'm happy with this.  It's basically asking the compiler to do the cast from high precision 64-bit int to less precision 53-bit *2^11 int in the right direction.

Reply:
Yes, UBSan is happy with it; tested internally with all the sanitizers, and with Debian's clang 7.0 with CFLAGS=-fsanitize=float-cast-overflow LDFLAGS=-fsanitize=float-cast-overflow.

Reply:
Does this condition still allow the case of `exponent.real` equal to `2**63` (because `(double)LONG_MAX` will be rounded up to `2**63`, assuming 64-bit longs)?

Reply:
If you look at the next bit of code, it's checking that int_exponent == exponent.real before using int_exponent. It's checking for a fast path for `complex ** int` (or float with an integer value) _after_ creating a complex out of `int`. I believe the new check has exactly the same effect while avoiding undefined behaviour, but the failure mode here is taking the slightly slower path for integer values that don't round-trip through the complex's 'real' double.

Reply:
Right, but that check occurs after the undefined behaviour has already taken place. If `exponent.real` is `0x1.0p+63`, the guard in the `if` condition will be true, and then we invoke undefined behaviour when we do `(long)exponent.real`.

Reply:
That's assuming that the conversion in `(double)LONG_MAX` rounds to nearest, of course. The C standard doesn't guarantee the rounding behaviour there (C11 6.3.1.5 says "the result is either the nearest higher or nearest lower representable value, chosen in an implementation-defined manner"), but round-to-nearest is the only behaviour I've seen happen in practice for integer-to-float conversions.

Reply:
Whoops, sorry; wrong section reference. That should be C11 §6.3.1.4p2, not C11 §6.3.1.5 (but the same quoted text).

Reply:
I think changing the second inequality to `exponent.real < -(double)LONG_MIN` would fix this corner case. Unlike `LONG_MAX`, the conversion of `LONG_MIN` to a double should be exact (assuming two's complement, no trap representation, etc., etc., but those are assumptions that we've been happy to make in many other places in the code).

Reply:
Yep, you're right. I didn't realise a double that's equal to LONG_MAX could end up being an invalid cast to long. I've added a test that triggered the behaviour, and was correctly caught by UBSan without the fix:

test_pow (__main__.ComplexTest) ... ../../cpython/Objects/complexobject.c:530:24: runtime error: 9.22337e+18 is outside the range of representable values of type 'long'

Reply:
Well, technically the double `0x1.0p+63` _isn't_ equal to `LONG_MAX`; it's equal to the result of converting `LONG_MAX` to a `double` - the error introduced in that conversion is exactly the cause of difficulties here.

Thanks for the fix! (BTW, I'm not Mark Shannon. :-))

Reply:
D'oh, I'm sorry, Mark! I know you're not Mark Shannon, I have no idea what I was thinking while I wrote that commit message.